proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+
+* Call arguments, e.g. `generic(repeat each value)`
+* Subscript arguments, e.g. `subscriptable[repeat each index]`
+* The elements of a tuple value, e.g. `(repeat each value)`
+* The elements of an array literal, e.g. `[repeat each value]`
+
+Pack expansion expressions can also appear in an expression statement at the top level of a brace statement. In this case, the semantics are the same as scalar expression statements; the expression is evaluated for its side effect and the results discarded.
+
+Note that pack expansion expressions can also reference _type_ pack parameters, as metatypes.
+
+**Capture:** A pack expansion expression _captures_ a value (or type) pack parameter the value (or type) pack parameter appears as a sub-expression without any intervening pack expansion expression.
+
+Furthermore, a pack expansion expression also captures all type parameter packs captured by the types of its captured value parameter packs.
+
+For example, say that `x` and `y` are both value parameter packs and `T` is a type parameter pack, and consider the pack expansion expression `repeat foo(each x, (each T).self, (repeat each y))`. This expression captures both the value parameter pack `x` and type parameter pack `T`, but it does not capture `y`, because `y` is captured by the inner pack expansion expression `repeat each y`. Additionally, if `x` has the type `Foo<U, (repeat each V)>`, then our expression captures `U`, but not `V`, because again, `V` is captured by the inner pack expansion type `repeat each V`.
+
+**Typing rules:** For a pack expansion expression to be well-typed, two conditions must hold:
+
+1. The types of all value parameter packs captured by a pack expansion expression must be related via same-shape requirements.
+
+2. After replacing all value parameter packs with non-pack parameters that have equivalent types, the pattern expression must be well-typed.
+
+Assuming the above hold, the type of a pack expansion expression is defined to be the pack expansion type whose pattern type is the type of the pattern expression.
+
+**Evaluation semantics:** At runtime, each value (or type) pack parameter receives a value (or type) pack, which is a concrete list of values (or types). The same-shape requirements guarantee that all value (and type) packs have the same length, call it `N`. The evaluation semantics are that for each successive `i` such that `0 ≤ i < N`, the pattern expression is evaluated after substituting each occurrence of a value (or type) pack parameter with the `i`th element of the value (or type) pack. The evaluation proceeds from left to right according to the usual evaluation order, and the sequence of results from each evaluation forms the argument list for the parent expression.
+
```

Comment:
A late-breaking question in second pitch thread was about how trailing closures will work with parameter packs. Seems reasonable to address particularly with the various "forward-scan" versus "backward-scan" rules of trailing closures that depend on Swift language mode.

Reply:
I've clarified in the proposal that parameter packs always use a forward scan for trailing closures.